### PR TITLE
Fix inferred argument names in `Module.simplify_signature/2`

### PR DESCRIPTION
**Pull request by whatyouhide**
Arguments like literals (e.g., `4`) were converted into variables like `int1` by `Module.simplify_signature/2`. This commit changes this behaviour by adding an index only if necessary (i.e., there are duplicates in the signature). For example:

```elixir
def foo([], 1, %{}, 2)
```

becomes

```elixir
def foo(list, int1, map, int2)
```

The code I added in this PR is not the cleanest code I ever wrote, but I couldn't come up with anything better. It works though :)

Let me know if I can do anything else, thanks!

### CODE REVIEW

**Comment by josevalim** [comment]
I think I have a better solution in mind.

Traverse the signature two times in exactly same way (using mapfoldl). The first one, you are interested in just the accumulator, to accumulate values using something like `{:collect, []}`. Then, the second time, you will use `{:generate, []}`.

For this to work, in simplify signature, you will call:

```elixir
defp simplify_signature(other, acc) when is_integer(other) do
  autogenerated(acc, :int)
end
```

where:

```elixir
defp autogenerated({:collect, acc}, key) do
  {:we_dont_care_about_this_yet, {:collect, [key|acc]}
end

defp autogenerated({:generate, acc}, key) do
  case previous_values(key, acc) do
    {i, acc} -> {{:"#{key}#{i}", [], Elixir}, {:generate, acc}}
    :none -> {{key, [], Elixir}, {:generate, acc}}
  end
end
```

This way, all the logic stays in autogenerated as long as you keep the info in the accumulator. You won't need to wrap with autogenerate and regular too. Thoughts?

**Comment by whatyouhide** [comment]
I don't remember who talked about @josevalim-driven-development at ElixirConfEU but boy was he right :). `push --force`'d a new version with your implementation almost verbatim. The code is a little bit cleaner (not sure the whole process of calling `simplify_signarure/2` *twice* for each arg would make sense to someone reading the code for the first time, but the code is indeed shorter and cleaner), so I think this approach is the best one.

All tests pass (I slightly changed the tests in `kernel/docs_test.exs`), but documentation is kind of giving me a hard time since I keep getting this error

```
** (ArgumentError) argument error
    :erlang.atom_to_binary("stat", :utf8)
    (elixir) lib/macro.ex:459: Macro.to_string/2
    (elixir) lib/enum.ex:1072: anonymous fn/4 in Enum.map_join/3
    (elixir) lib/enum.ex:1383: Enum."-reduce/3-lists^foldl/2-0-"/3
    (elixir) lib/enum.ex:1071: Enum.map_join/3
    (elixir) lib/macro.ex:672: Macro.call_to_string_with_args/3
    (elixir) lib/macro.ex:576: Macro.to_string/2
    lib/ex_doc/retriever.ex:161: ExDoc.Retriever.get_function/5
```

when I run `make docs`. I'm not sure it's related with my changes though (albeit probably it is) since it's a bit cryptic. I'll wait for some advice here :)

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -569,6 +603,48 @@ defmodule Module do
   defp do_camelcase_to_underscore(<<>>, acc),
     do: acc
 
+  defp resolve_signature_duplicates(signatures) do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: We no longer need this, right?

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -528,37 +526,73 @@ defmodule Module do
 
   # Simplify signatures to be stored in docs
 
-  defp simplify_signature({:\\, _, [left, right ]}, i) do
-    {:\\, [], [simplify_signature(left, i), right]}
+  defp simplify_signature(signature) do
+    folder = fn(el, acc) -> simplify_signature(el, acc) end
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: `&simplify_signature/2`

File: lib/elixir/lib/module.ex [hunk 3]
```diff
@@ -528,37 +526,73 @@ defmodule Module do
...
     case Atom.to_string(var) do
-      "_" <> rest -> {String.to_atom(rest), [], Elixir}
-      _           -> {var, [], nil}
+      "_" <> rest -> {{String.to_atom(rest), [], Elixir}, acc}
+      _           -> {{var, [], nil}, acc}
+    end
+  end
+
+  defp simplify_signature({:%, _, [left, _]}, acc) when is_atom(left) do
+    struct_name = camelcase_to_underscore(List.last(split(left)))
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: The bug is coming from here. `camelcase_to_underscore` is returning a string which later on may end-up being used as key. We need to convert it to an atom before calling autogenerated.

File: lib/elixir/lib/regex.ex [hunk 4]
```diff
@@ -85,6 +85,8 @@ defmodule Regex do
     defexception message: "regex could not be compiled"
   end
 
+  def foo(1, 2, %{}, 4, %{}, protobufferizer), do: :foo
```
Comment (josevalim) [inline lib/elixir/lib/regex.ex]: :scissors: :)

**Comment by whatyouhide** [comment]
This PR is a mess, I have to stop working so late at night :). Fixes are coming!

**Comment by whatyouhide** [comment]
Fixes came! Again, sorry for the mess. I'm very happy for this PR though, it adds functionality that I personally **love** :). I'm sure @eksperimental will be happy as well, he often takes care of docs and of function signatures in particular lately IIRC.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**